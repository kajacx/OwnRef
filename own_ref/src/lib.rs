mod macros;
mod structs;

pub use structs::OwnRef;
pub use structs::OwnRefMut;
