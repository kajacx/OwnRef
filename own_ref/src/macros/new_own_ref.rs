#[macro_export]
macro_rules! new_own_ref {
    (mut $ref_name:ident, $data_variable:ident) => {
        let _own_ref_phantom_scope = ();
        let mut $ref_name = unsafe {
            // SAFETY:
            // We know the OwnRef will not outlive the current stack,
            // because of the _own_ref_phantom_scope lifetime.
            // Also, the data cannot cannot be used after "moving it" into the OwnRef
            // because we forget it immediately.
            // Finally, the data will not be deallocated twice,
            // because we forget it and not drop it.
            #[allow(deprecated)]
            own_ref::internals::new_own_ref_mut(
                &mut $data_variable,
                own_ref::internals::lifetime_of(&_own_ref_phantom_scope),
            )
        };
        std::mem::forget($data_variable);
    };
    ($ref_name:ident, $data_variable:ident) => {
        let _own_ref_phantom_scope = ();
        let $ref_name = unsafe {
            // SAFETY:
            // We know the OwnRef will not outlive the current stack,
            // because of the _own_ref_phantom_scope lifetime.
            // Also, the data cannot cannot be used after "moving it" into the OwnRef
            // because we forget it immediately.
            // Finally, the data will not be deallocated twice,
            // because we forget it and not drop it.
            #[allow(deprecated)]
            own_ref::internals::new_own_ref(
                &$data_variable,
                own_ref::internals::lifetime_of(&_own_ref_phantom_scope),
            )
        };
        std::mem::forget($data_variable);
    };
}
