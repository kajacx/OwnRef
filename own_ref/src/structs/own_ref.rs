use std::marker::PhantomData;
use std::ops::Deref;
use std::ptr::NonNull;

pub struct OwnRef<'a, T> {
    pointer: NonNull<T>,
    _phantom_lifetime: PhantomData<&'a ()>,
    _phantom_value: PhantomData<T>,
}

impl<'a, T> OwnRef<'a, T> {
    pub(crate) unsafe fn new<'b>(reference: &'b T, _phantom_lifetime: PhantomData<&'a ()>) -> Self {
        Self {
            pointer: NonNull::from(reference),
            _phantom_lifetime,
            _phantom_value: PhantomData,
        }
    }

    pub(crate) fn from_pointer(
        pointer: NonNull<T>,
        _phantom_lifetime: PhantomData<&'a ()>,
    ) -> Self {
        Self {
            pointer,
            _phantom_lifetime,
            _phantom_value: PhantomData,
        }
    }

    pub fn get(&self) -> &T {
        unsafe { self.pointer.as_ref() }
    }

    pub fn take(self) -> T {
        let result = unsafe { std::ptr::read(self.pointer.as_ptr()) };
        std::mem::forget(self); // do not run destructor
        result
    }
}

impl<'a, T> Deref for OwnRef<'a, T> {
    type Target = T;

    fn deref(&self) -> &Self::Target {
        self.get()
    }
}

impl<T> Drop for OwnRef<'_, T> {
    fn drop(&mut self) {
        let _drop = unsafe { std::ptr::read(self.pointer.as_ptr()) };
    }
}
