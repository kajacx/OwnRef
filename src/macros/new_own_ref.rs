#[macro_export]
macro_rules! new_own_ref {
    ($ref_name:ident, $data_variable:ident) => {
        let _phantom = ();
        let $ref_name = unsafe {
            // SAFETY:
            // We know the OwnRef will not outlive the current stack,
            // because of the _phantom lifetime.
            // Also, the data cannot cannot be used after "moving it" into the OwnRef
            // because we forget it immediately.
            // Finally, the data will not be deallocated twice,
            // because we forget it and not drop it.
            crate::structs::own_ref::OwnRef::new(&mut $data_variable, &_phantom)
        };
        std::mem::forget($data_variable);
    };
}

pub(crate) use new_own_ref;
