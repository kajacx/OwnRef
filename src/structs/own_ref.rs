use std::marker::PhantomData;
use std::ops::Deref;
use std::ptr::NonNull;

pub struct OwnRef<'a, T> {
    pointer: NonNull<T>,
    _phantom_lifetime: PhantomData<&'a ()>,
    _phantom_value: PhantomData<T>,
}

impl<'a, T> OwnRef<'a, T> {
    pub unsafe fn new<'b>(reference: &'b T, _phantom_lifetime: &'a ()) -> Self {
        Self {
            pointer: NonNull::from(reference),
            _phantom_lifetime: PhantomData,
            _phantom_value: PhantomData,
        }
    }

    pub fn get(&self) -> &T {
        unsafe { self.pointer.as_ref() }
    }

    pub fn take(self) -> T {
        let result = unsafe { std::ptr::read(self.pointer.as_ptr()) };
        std::mem::forget(self); // do not run destructor
        result
    }
}

impl<'a, T> Deref for OwnRef<'a, T> {
    type Target = T;

    fn deref(&self) -> &Self::Target {
        self.get()
    }
}

impl<T> Drop for OwnRef<'_, T> {
    fn drop(&mut self) {
        let _drop = unsafe { std::ptr::read(self.pointer.as_ptr()) };
    }
}
