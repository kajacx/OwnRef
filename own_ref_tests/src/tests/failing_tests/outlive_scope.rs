use own_ref::new_own_ref;

fn main() {
    let external;

    if 1 > 2 {
        let scoped = "Hello".to_owned();
        new_own_ref!(as_own_ref, scoped);
        external = as_own_ref;
    } else {
        return;
    }

    println!("Oh no, out of scope memory: {}", external.as_str());
}
