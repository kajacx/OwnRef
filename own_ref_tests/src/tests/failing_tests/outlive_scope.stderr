error[E0597]: `_own_ref_phantom_scope` does not live long enough
  --> src/tests/failing_tests/outlive_scope.rs:8:9
   |
8  |         new_own_ref!(as_own_ref, scoped);
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ borrowed value does not live long enough
9  |         external = as_own_ref;
10 |     } else {
   |     - `_own_ref_phantom_scope` dropped here while still borrowed
...
14 |     println!("Oh no, out of scope memory: {}", external.as_str());
   |                                                ----------------- borrow later used here
   |
   = note: this error originates in the macro `new_own_ref` (in Nightly builds, run with -Z macro-backtrace for more info)
